doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,359 @@
+# Overview
+
+People want to use Materialize to maintain SQL queries over recent histories of their append-only
+data. This is because persisting append-only data grows without bound, making rehydration both
+inefficient and expensive. Giving users more control into how and when to expire their data makes
+Materialize applicable to broader set of use cases and customers.
+
+# Goals
+
+This design doc proposes supporting a limited set of data transformations during ingestion made
+available to users by extending the syntax of the `CREATE SOURCE` statement. The extension has the
+form of a new `TRANSFORM USING` keyword which is paramterized by a SQL query describing the
+transformation.
+
+The goal are to allow users to:
+
+* Thin down their data through expiration, by using a temporal filter in their transformation.
+* More generally filter and transform their data before peristing them through SQL expressions.
+
+### Non-goals
+
+- Express existing envelopes and decoders in terms of a transformation query instead of custom code.
+  It seems possible but it’s out of scope for the initial implementation of this feature.
+- Allow arbitrary computation in the transformation queries.
+- Define syntax for both transforming and breaking up a source into subsources.
+
+## Feature description
+
+### Syntax
+
+The current ingestion syntax can be viewed as a series of transformations that are applied one after
+the other. For example users can add a `FORMAT` transformation to do format decoding on top of the
+raw output and an `ENVELOPE` transformation to interpret a certain types of envelopes after format
+decoding.
+
+This proposal introduces an additional transformation whose general form is `TRANSFORM USING
+(<query>)`. Semantically this transformation could be applied at before or after any of the existing
+transformations (i.e pre/post format and pre/post envelope) we will limit the allowed positions for
+the initial release of this feature to effectively post envelope. In other words, the schema that
+this query should expect to operate on will be the same schema that the source would have produced
+without it.
+
+**Example: Decode JSON and keep only certain rows**
+
+```sql
+CREATE SOURCE sales_staff
+FROM KAFKA CONNECTION ...
+FORMAT BYTES
+TRANSFORM USING (
+    SELECT
+        (data->>'id')::int AS id,
+        (data->>'name')::text AS name,
+    FROM
+        (SELECT CONVERT_FROM(data, 'utf8')::jsonb as data FROM sales_staff)
+    WHERE
+        (data->>'department')::text == 'sales'
+);
+```
```

Comment:
As an example of applying the above suggestion, this SQL query becomes:

```suggestion
```sql
CREATE SOURCE sales_staff
FROM KAFKA CONNECTION ...
FORMAT BYTES
TRANSFORM (
    SELECT
        (data->>'id')::int AS id,
        (data->>'name')::text AS name,
    FROM
        (SELECT CONVERT_FROM(data, 'utf8')::jsonb as data FROM sales_staff)
    WHERE
        (data->>'department')::text == 'sales'
);
```
```